### PR TITLE
Export test refactor

### PR DESCRIPTION
**Pull request by itsmurugappan**
## Description

Refactor kn service export test to use helper functions from serving repo.

First and third commit - Has only vendor changes
Second commit - Has the test refactor changes

### CODE REVIEW

File: lib/test/service.go [hunk 1]
```diff
@@ -15,18 +15,28 @@
...
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/ptr"
 	pkgtest "knative.dev/pkg/test"
 	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+	servingtest "knative.dev/serving/pkg/testing/v1"
 
 	"knative.dev/client/pkg/util"
 )
 
+type ExpectedServiceListOption func(*servingv1.ServiceList)
```
[the remark this answers is not in the document]
Comment (navidshaikh) [reply lib/test/service.go:36]: @itsmurugappan : We should add the description to all these exported types as they're in `lib` and are expected to be consumed by plugins etc, it helps in the editor.

File: lib/test/service.go [hunk 2]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+
+	for _, fn := range options {
+		fn(knExport)
+	}
+
+	return knExport
+}
+
+//Cfg builds servingv1.Configuration with the options provided
+func Cfg(co ...servingtest.ConfigOption) *servingv1.ConfigurationSpec {
```
Comment (navidshaikh) [inline lib/test/service.go:204]: how about renaming this to `ServiceConfig` ?

File: lib/test/service.go [hunk 3]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+
+var revisionSpec = servingv1.RevisionSpec{
+	PodSpec: corev1.PodSpec{
+		Containers: []corev1.Container{{
+			Image: pkgtest.ImagePath("helloworld"),
+		}},
+		EnableServiceLinks: ptr.Bool(false),
+	},
+	TimeoutSeconds: ptr.Int64(300),
+}
```
Comment (navidshaikh) [inline lib/test/service.go:227]: move the var at the top of the file ?

File: lib/test/service.go [hunk 4]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+
+var revisionSpec = servingv1.RevisionSpec{
+	PodSpec: corev1.PodSpec{
+		Containers: []corev1.Container{{
+			Image: pkgtest.ImagePath("helloworld"),
+		}},
+		EnableServiceLinks: ptr.Bool(false),
+	},
+	TimeoutSeconds: ptr.Int64(300),
```
Comment (navidshaikh) [inline lib/test/service.go:226]: should use `DefaultRevisionTimeoutSeconds` constant from `knative.dev/serving/pkg/apis/config/defaults.go`

File: lib/test/service.go [hunk 5]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+}
+
+//GetSvcWithOptions returns ksvc with options provided
```
Comment (navidshaikh) [inline lib/test/service.go:229]:
```suggestion
// GetSvcWithOptions returns ksvc with options provided
```

File: lib/test/service.go [hunk 6]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+	svc.TypeMeta = metav1.TypeMeta{
+		Kind:       "Service",
+		APIVersion: "serving.knative.dev/v1",
+	}
+	svc.Spec.Template.Spec.Containers[0].Resources = corev1.ResourceRequirements{}
+	return svc
+}
+
+//WithTrafficSplit adds route to ksvc
+func WithTrafficSplit(revisions []string, percentages []int, tags []string) servingtest.ServiceOption {
```
Comment (navidshaikh) [inline lib/test/service.go:241]:
```suggestion
func WithTrafficSpec(revisions []string, percentages []int, tags []string) servingtest.ServiceOption {
```

File: lib/test/service.go [hunk 7]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+			}
+		}
+		svc.Spec.RouteSpec = servingv1.RouteSpec{
+			Traffic: trafficTargets,
+		}
+	}
+}
+
+//GetRev returns Revision object with the options provided
+func GetRev(name string, options ...servingtest.RevisionOption) servingv1.Revision {
```
Comment (navidshaikh) [inline lib/test/service.go:265]: above, `Cfg` is used to build the configuration and here `GetRev`; all these builder utils should follow the same pattern, IMO we should avoid short names as well. 
wdyt about `BuildRevision` / `BuildConfiguration`?
I'd also avoid `Get` verb here as that denote to client GET call.

File: lib/test/service.go [hunk 8]
```diff
@@ -134,3 +144,171 @@ func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string
...
+		Kind:       "Revision",
+		APIVersion: "serving.knative.dev/v1",
+	}
+	rev.Spec.PodSpec.Containers[0].Name = "user-container"
+	rev.ObjectMeta.SelfLink = ""
+	rev.ObjectMeta.Namespace = ""
+	rev.ObjectMeta.UID = ""
+	rev.ObjectMeta.Generation = int64(0)
+	rev.Spec.PodSpec.Containers[0].Resources = corev1.ResourceRequirements{}
+	return *rev
```
Comment (navidshaikh) [inline lib/test/service.go:277]: may pass the pointer instead ? (like done for Configuration above)

**Review by navidshaikh — CHANGES_REQUESTED** [review]
Thanks for promoting these utils to `lib`!

There was a fix landed for the failure in e2e (against master), looks like that's lost during refactor ?
```
        --- knExport
        +++ &actSvc
        [...]

        - 						EnableServiceLinks:        nil,
        + 						EnableServiceLinks:        &false,
        [...]
```

**Review by maximilien — APPROVED** [review]
LGTM

Wonder if these vendor changes could have been part of a separate commit/PR?. Never easy to review... Anyhow.

**Review by navidshaikh — CHANGES_REQUESTED** [review]
a few places left in the code which can be aligned with the latest improvements made to the PR, i.e. to use full name `Revision` over `Rev`. Also, if the builder function takes a single object of a type, IMO we should consider using singular name for the function, for example `WithRevision`.

**Comment by maximilien** [comment]
@itsmurugappan you need to rebase to pickup latest deps

**Review by navidshaikh — APPROVED** [review]
/lgtm
/approve

Thank you @itsmurugappan for the refactor! 
There are a few other parts for example the flags definition etc that we'd need regroup in a better way, lets talk about it in upcoming WG call.

**Comment by navidshaikh** [comment]
@itsmurugappan : also please send another with an entry in CHANGELOG :-)

**Comment by itsmurugappan** [comment]
sure @navidshaikh can include the changelog and other changes in the pr for bug fix #1087

**Pull request merged**